### ISSUE TITLE
Extend treatment on Unicode and/or its security considerations

### ISSUE THREAD

**Issue by awwright** [json-schema-org/json-schema-spec#215]
Unicode is a complex technology that probably nobody will ever fully understand. But we should add a few notes on common implementation conserns, especially security considerations.

Also consider the behavior of applications that use e.g. UTF-16:

~~~
> '🐲' .length // U+1F432
< 2
~~~

**Comment by brettz9** [comment]
Drawing out some implications of your example... With `maxLength`/`minLength`, the validation spec states these refer to the "number of its characters as defined by RFC 7159." (the JSON spec)

The latter, however, while referring to Unicode "characters" as being escaped as UTF-16, also states, "implementations might return different values for the length of a string value", so it would probably help to be more clear on what the intention is here in deferring to the JSON spec.

For example, to enforce the string length is no longer than in your example, should `maxLength` be 1 or 2? I don't think the current spec is actually very clear on this.

**Comment by akuckartz** [comment]
I agree that it makes sense to think about security aspects of Unicode. But these aspects are not specific to JSON Schema. A separate document might make sense which can be developed by a broader community (including JSON-LD supporters for example).

**Comment by brettz9** [comment]
In the case of `maxLength` and `minLength`, if one mistakenly relies on them, these are JSON Schema-specific issues. But again, I don't think the behavior is clearly spec'd.

**Comment by epoberezkin** [comment]
> should maxLength be 1 or 2

@brettz9 there are tests that require it's 1

**Comment by brettz9** [comment]
Sure, @epoberezkin , but the text ought to be clarified regardless.

**Comment by micolous** [comment]
One should also consider combining characters (and [normalisation forms](https://en.wikipedia.org/wiki/Unicode_equivalence#Normal_forms)) and zero-width joiner combinations (where [platforms have included custom rules](https://blog.emojipedia.org/ninja-cat-the-windows-only-emoji/)). Both of these will change over time with the Unicode standard.

So you could length as:

* bytes (which can vary depending on whether you use UTF-8, UTF-16, UTF-32, etc., as well as how the string is composed, but it's also the simplest to implement)
* codepoints (fairly consistent, but codepoints vary in byte length and also vary depending on composition)
* characters (for UTF-16 and UTF-32, can be different from codepoints)
* graphemes (composition doesn't matter, interpretation evolves with the Unicode standard, but it's closer to the human understanding of "this is a character" in *many, but not all* languages)

You could hand-wave away composition and say "it will be counted based on how the data is transmitted", with the idea that a client *should* [convert to NFC](https://en.wikipedia.org/wiki/Unicode_equivalence#Normal_forms).

> there are tests that require it's 1

@epoberezkin Using that example, that might be counting in codepoints or graphemes.

A better example is: "👩🏾‍🚀" (woman emoji + skin tone + ZWJ + rocket emoji). You could count that as:

* bytes: 15 (in UTF-8)
* JavaScript: 7 (UTF-16 characters)
* codepoints: 4
* graphemes: 1

Which one is the correct answer for JSON Schema's `minLength`/`maxLength`?

And does everyone implement it that way? 😅

**Comment by gregsdennis** [comment]
The current tests check for codepoints.

[RFC 8259, Section 8.3](https://www.rfc-editor.org/rfc/rfc8259.html#section-8.3), mentions "Unicode code units".  It's reasonable that regarding Unicode as such applies to the rest of the specification where Unicode is mentioned.

I imagine we could make an explicit declaration in [Validation, Section 4.1](https://json-schema.org/draft/2020-12/json-schema-validation#section-4.1).

**Comment by micolous** [comment]
Going back and answering my own question: I had a look at the test suite, and it looks like it is actually implemented in terms of graphemes since https://github.com/json-schema-org/JSON-Schema-Test-Suite/pull/52:

For `maxLength: 2`:

https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/69136952196a63a7553803935feaeaec57a48420/tests/draft4/maxLength.json#L26-L30

For `minLength: 2`:

https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/69136952196a63a7553803935feaeaec57a48420/tests/draft4/minLength.json#L26-L30

https://github.com/json-schema-org/JSON-Schema-Test-Suite/pull/710 updated `draft-next` to use the word “grapheme”.

**Comment by gregsdennis** [comment]
It appears we already have a recommendation about Unicode in regular expressions in [Core 6.4](https://json-schema.org/draft/2020-12/json-schema-core#section-6.4):

> Furthermore, given the high disparity in regular expression constructs support, schema authors SHOULD limit themselves to the following regular expression tokens:
> - individual Unicode characters, as defined by the JSON specification;
> - ...

[RFC 8259, section 8.3](https://datatracker.ietf.org/doc/html/rfc8259#section-8.3) states:

> Implementations that transform the textual representation into sequences of Unicode code units and then perform the comparison numerically, code unit by code unit, are interoperable in the sense that implementations will agree in all cases on equality or inequality of two strings.

We also have in [Core 4.2.1](https://json-schema.org/draft/2020-12/json-schema-core#section-4.2.1) that strings are defined as `A string of Unicode code points, from the JSON "string" value`.

I feel this is sufficient, but I'm open to any clarification suggestions.

**Comment by micolous** [comment]
> It appears we already have a recommendation about Unicode in regular expressions in [Core 6.4](https://json-schema.org/draft/2020-12/json-schema-core#section-6.4):
> 
> > Furthermore, given the high disparity in regular expression constructs support, schema authors SHOULD limit themselves to the following regular expression tokens:
> > 
> > * individual Unicode characters, as defined by the JSON specification;

[Errata to RFC 8259](https://www.rfc-editor.org/errata/eid7603) takes issue with the term "Unicode characters":

> Section 1 says:
>
>>  A string is a sequence of zero or more Unicode characters [UNICODE].
>
> It should say:
>
>>  A string is a sequence of zero or more Unicode code points [UNICODE].
>
> Notes:
>
> Surrogate code points are not Unicode characters, as explained here: https://www.unicode.org/glossary/#surrogate_character
>
> However, a surrogate code point outside of a surrogate pair is allowed in JSON strings both in escaped and unescaped forms according to the ABNF grammar in section 7 and the warning in section 8.2, despite an UTF-8 incompatibility for the unescaped form. In addition, the original text contradicts ECMA-404 section 9, which states: "A string is a sequence of Unicode code points wrapped with quotation marks (U+0022). All code points may be placed within the quotation marks except for the code points that must be escaped: quotation mark (U+0022), reverse solidus (U+005C), and the control characters U+0000 to U+001F. " 

This would suggest that we should follow the ECMA-404 definition of a string, which is defined in terms of [code points](https://www.unicode.org/glossary/#code_point).

Therefore, "👩🏾‍🚀" (woman emoji + skin tone + ZWJ + rocket emoji) would be of length 4.

> [RFC 8259, section 8.3](https://datatracker.ietf.org/doc/html/rfc8259#section-8.3) states:
> 
> > Implementations that transform the textual representation into sequences of Unicode code units and then perform the comparison numerically, code unit by code unit, are interoperable in the sense that implementations will agree in all cases on equality or inequality of two strings.

That explains how to test for equality, and that it should be based on [code units](https://www.unicode.org/glossary/#code_unit) (= bytes, when encoded as UTF-8, or `uint16_t` when encoded as UTF-16), rather than code points or graphemes.

Therefore, "👩🏾‍🚀" (woman emoji + skin tone + ZWJ + rocket emoji) would be of length 15 (in UTF-8) or 7 (in UTF-16).

> I feel this is sufficient, but I'm open to any clarification suggestions.

In [Section 6.3 (Validation keywords for strings)](https://json-schema.org/draft/2020-12/json-schema-validation#section-6.3), change:

> The length of a string instance is defined as *the number of its characters as defined by [RFC 8259](https://json-schema.org/draft/2020-12/json-schema-validation#RFC8259)*

to one of:

* ...the number of its [code points](https://www.unicode.org/glossary/#code_point), as defined by Unicode.
* ...the number of its [code units](https://www.unicode.org/glossary/#code_unit), as defined by Unicode.
* ...the number of its [graphemes](https://www.unicode.org/glossary/#grapheme), as defined by Unicode.

And note what the length of "👩🏾‍🚀" should be interpreted as (and if "code unit", then note what encoding it should use, eg: UTF-8).

**Comment by jdesrosiers** [comment]
I agree that we should update the spec to make this more clear. I believe counting the number of "code points" is correct. We say, "The length of a string instance is defined as the number of its characters as defined by [RFC 8259](https://json-schema.org/draft/2020-12/json-schema-validation#RFC8259)". RFC 8259 defines a string as, "a sequence of zero or more Unicode characters". Therefore, we're counting unicode characters. A unicode character is identified by a code point, so that's the same as saying we're counting code points.

If we were to say we're counting code units, that would break the poop emoji tests we've had forever and would be a breaking change for every existing implementation. Even if there's a compelling argument that that's technically what the spec says, we can't do that. We would need to change the spec to reflect reality.

Unfortunately, I think counting code points is the least useful way for `minLength`/`maxLength` to behave. Counting code units (bytes) is useful if you have fixed memory the string can fit into. Counting graphemes is useful if you have fixed number of characters to display the string in a UI. Counting code points means `minLength`/`maxLength` isn't reliable for either of those uses and I can't think of a case where counting code points makes sense.

**Comment by gregsdennis** [comment]
@jdesrosiers I'm not sure where your recommendation lies.  You talk about the poop emoji tests (which use code points) that would break existing tests, but then you talk about how that's the least useful way to measure.

So do we stick with code points or move to one of the more "useful" measurements?

**Comment by jdesrosiers** [comment]
Sorry. Ignore the last paragraph. I was just complaining. Despite its issues, I don't think we want to make breaking changes to these keywords. I interpret the spec to mean we're counting code points which is consistent with the poop emoji tests and what I believe has been the accepted interpretation for many years.

My suggestion is that we add some note to the spec to make it more clear that we're counting code-points so readers don't need to do a deep dive on unicode figure out what that lines means.

**Comment by micolous** [comment]
> RFC 8259 defines a string as, "a sequence of zero or more Unicode characters". Therefore, we're counting unicode characters. A unicode character is identified by a code point, so that's the same as saying we're counting code points.

[As I noted above](https://github.com/json-schema-org/json-schema-spec/issues/215#issuecomment-2955709660), the phrase "Unicode character" is ambiguous, and there's already errata to RFC 8259 to this effect.

> I interpret the spec to mean we're counting code points which is consistent with the poop emoji tests and what I believe has been the accepted interpretation for many years.

[As I also noted above, the poop emoji test doesn't actually solve whether it's code points or graphemes](https://github.com/json-schema-org/json-schema-spec/issues/215#issuecomment-2760249427), because it has a single code point, which is why I gave the astronaut with skin tone example (which is a single grapheme with multiple code points).

[The JSON Schema reference material](https://json-schema.org/understanding-json-schema/reference/string) is even more ambiguous, because it only uses basic Latin characters that are all representable with a single byte in UTF-8, and doesn't reference RFC 8259 at all.

> My suggestion is that we add some note to the spec to make it more clear that we're counting code-points so readers don't need to do a deep dive on unicode figure out what that lines means.

For what it's worth, I think that counting code points is probably the simplest answer, provided it notes some caveats:

* "Unicode characters" is ambiguous
* code points _aren't_ the same as graphemes
* `String.length()` can do different things depending on your programming language (and sometimes, how it is configured)

While it feels a bit silly to use emoji as examples, skin tone modifiers and flag emoji are fairly well-known applications of combining and joining characters which are fairly easy to explain (eg: "🇦🇺" [is 2 code points](https://www.unicode.org/reports/tr51/tr51-27.html#Flags), "🏳️‍🌈" is 4,  "🏴󠁧󠁢󠁳󠁣󠁴󠁿" [is 7](https://www.unicode.org/reports/tr51/tr51-27.html#Sample_Valid_Emoji_Tag_Sequences)).

There are (written) languages which rely on combining characters in more complex ways, but they are harder to explain, so I think it is sufficient to simply acknowledge that they exist, and the limitations of what `minLength` and `maxLength` can _practically_ express.

If you're writing a JSON Schema document (or something based on it, like OpenAPI), you need to be aware of these issues, and that you can't expect a 1:1 mapping between your UI/UX requirements and API.

There may be room for a JSON Schema extension which expresses constraints in graphemes, which could map more closely to UI/UX requirements. However, evaluating those attributes depends on the CLDR, which is a large and moving target, not to mention some platform specific extensions.

**Comment by jdesrosiers** [comment]
As I'm rereading my previous comment, a lot of that doesn't make sense. Sorry for adding confusion. However, the conclusion of counting code points does make sense.

> the phrase "Unicode character" is ambiguous, and there's already errata to RFC 8259 to this effect.

Sorry, I missed that errata. But, that only seems to strengthen the conclusion that we should be counting code points. I agree that the term "unicode character" isn't precise. If were using that term in the spec, we should fix it, but I don't think we are.

> As I also noted above, the poop emoji test doesn't actually solve whether it's code points or graphemes

My point in mentioning the poop emoji was that whatever interpretation we settle on MUST be consistent with the poop emoji test. It was understood that the poop emoji test is not sufficient to disambiguate how the length should be counted. We should add the astronaut example as an additional test when this is settled.

> For what it's worth, I think that counting code points is probably the simplest answer, provided it notes some caveats:
>
> - "Unicode characters" is ambiguous
> - code points aren't the same as graphemes
> - `String.length()` can do different things depending on your programming language (and sometimes, how it is configured)

I agree to all of those points, but I think that's starting to sound more like a tutorial than a specification. I think, for the spec, it's sufficient to say that the length calculated by `minLength` and `maxLength` is the number of code points in the string. It's out of scope for the spec to explain the difference between code points and graphemes. However, it would be helpful for the documentation on the website to explain these things.

> There are (written) languages which rely on combining characters in more complex ways, but they are harder to explain, so I think it is sufficient to simply acknowledge that they exist, and the limitations of what `minLength` and `maxLength` can _practically_ express.

Agreed. I think that would be a good addition to the documentation on the website.

> There may be room for a JSON Schema extension which expresses constraints in graphemes, which could map more closely to UI/UX requirements.

My thoughts exactly!
